crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -778,11 +809,147 @@ impl<'vm, 'state, 'a, 'b, S: StorageView> ExecutionContext<'vm, 'state, 'a, 'b,
+    use move_core_types::vm_status::StatusCode;
+
+    let storage = session.get_resolver().storage();
+    match storage.get_package(&package_id) {
+        Ok(Some(package)) => Ok(package),
+        Ok(None) => Err(
+            PartialVMError::new(StatusCode::LINKER_ERROR)
+                .with_message(format!("Cannot find link context {package_id} in store"))
+                .finish(Location::Undefined),
+        ),
+        Err(err) => Err(
+            PartialVMError::new(StatusCode::LINKER_ERROR)
+                .with_message(format!("Error loading link context {package_id} from store: {err}"))
+                .finish(Location::Undefined),
+        ),
+    }
+}
+
+/// Load `type_tag` to get a `Type` in the provided `session`.  `session`'s linkage context will be
+/// reset after this operation, because during the operation, it may change when loading a struct.
+pub fn load_type<'vm, 'state, S: StorageView>(
+    session: &mut Session<'state, 'vm, LinkageView<'state, S>>,
+    type_tag: &TypeTag,
+) -> VMResult<Type> {
+    use move_binary_format::errors::PartialVMError;
+    use move_core_types::vm_status::StatusCode;
+
+    fn load_type_<'vm, 'state, S: StorageView>(
+        session: &mut Session<'state, 'vm, LinkageView<'state, S>>,
+        type_tag: &TypeTag,
+    ) -> VMResult<Type> {
+        Ok(match type_tag {
+            TypeTag::Bool => Type::Bool,
+            TypeTag::U8 => Type::U8,
+            TypeTag::U16 => Type::U16,
+            TypeTag::U32 => Type::U32,
+            TypeTag::U64 => Type::U64,
+            TypeTag::U128 => Type::U128,
+            TypeTag::U256 => Type::U256,
+            TypeTag::Address => Type::Address,
+            TypeTag::Signer => Type::Signer,
+
+            TypeTag::Vector(inner) => Type::Vector(Box::new(load_type_(session, inner)?)),
+            TypeTag::Struct(struct_tag) => {
+                let StructTag {
+                    address,
+                    module,
+                    name,
+                    type_params,
+                } = struct_tag.as_ref();
+
+                // Load the package that the struct is defined in, in storage
+                let defining_id = ObjectID::from_address(*address);
+                let package = package_for_linkage(session, defining_id)?;
+                let original_address = AccountAddress::from(package.original_package_id());
+                let runtime_id = ModuleId::new(original_address, module.clone());
+
+                // Set the defining package as the link context on the session while loading the
+                // struct
+                set_linkage(session, &package).map_err(|e| {
+                    PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                        .with_message(e.to_string())
+                        .finish(Location::Undefined)
+                })?;
+
+                let (idx, struct_type) = session.load_struct(&runtime_id, name)?;
+
```

Comment:
Nit (and carry over from the draft PR). To me it would be more intuitive to reason about setting/resetting context in this recursive function if it was reset here rather (and paired up with `set_linkage`) than once in line 944 after a bunch of sets.

Reply:
Sure, I can make that change.